apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1065 @@
+    'nvidia-cuda-runtime==12.9.79; platform_system == "Linux" and platform_machine == "x86_64" and "nvidia :: cuda_version_lower_bound :: 12.9" in variant_properties',
+]
+```
+
+</figure>
+
+There are three environment markers available for use: `variant_properties`
+corresponding to the set of the variant properties that the wheel was built for,
+`variant_features` combining them into enabled features (i.e. features that have
+at least one value), and `variant_namespaces` doing the same for namespaces.
+There has also been a discussion of adding a marker to match variant label
+specifically, i.e. `variant =​= "cu129"`, but so far there did not seem
+to emerge a really good use case for it, and it was deemed unnecessarily risky,
+as building the same variant with a different label would cause the dependency
+to stop applying.
+
+The discussion around environment markers surfaced more interesting problems.
+For a start, it was pointed out that there could be two different types
+of markers: markers matching the properties of the installed wheel (as outlined
+previously), and markers matching the current system state. For example,
+the latter could be used to install the CUDA plugin if CUDA is available without
+actually having to publish a separate CUDA variant of the main package (think
+JAX). However, this is a recent development and it has not been pursued yet.
+
+It was also pointed out that technically these kind of dependencies can create
+potential conflicts. For example, since technically a wheel can have multiple
+values for a property, you could end up pulling in two or three versions
+of `nvidia-cuda-runtime` simultaneously. This particularly causes problems
+when creating universal lockfiles — since they need to account for all valid
+combinations of environment markers.
+
+## Let's talk about security
+
+Probably the biggest problem posed by wheel variants are their security
+implications. Regular wheels are installable entirely without executing
+third-party code — there are no scripts involved. On the other hand,
+when installing variant wheels, one has to execute third-party code
+from provider plugins. This cannot be avoided without sacrificing
+the flexibility and user-friendliness.
+
+<div style={{ textAlign: "center" }}>
+<figure style={{width: 'auto', margin: '0 2em', display: 'inline-block', verticalAlign: 'top'}}>
+  <img src="/posts/python-wheels-from-tags-to-variants/variant-risks.png"
+    width="697" height="373" alt="A diagram showing the three main attack
+vectors at the three additional components of variant wheel's supply chain.
+Firstly, the variant wheel can be attacked directly by adding a malicious provider.
+Secondly, a provider can be attacked by adding malicious code to it. Thirdly,
+one of its dependencies can be attacked in the same manner." />
+  <figcaption>Fig. 10. Variant wheel supply chain attack vectors</figcaption>
+</figure>
+</div>
```

Comment:
I think I'd like this much better if it showed something like this:
```
Executed at package install time (may have elevated permissions)   |   Executed at time of usage of package/application

Installer + its deps                     | wheel contents
```

going to:

```
Executed at package install time (may have elevated permissions)   |   Executed at time of usage of package/application

Installer + its deps                     | wheel contents--> (arrow to variant plugin)
                                         | 
variant plugin + its deps                |   
```